apps/email/js/cards/cmp/autocomplete.js

```diff
@@ -0,0 +1,476 @@
+     */
+    match: function() {
+      if (!this.query || !this.dataSource) {
+        this.list.innerHTML = '';
+        this.hideList();
+        return;
+      }
+
+      if (!this.timedQueryId) {
+        // The choice of delay tries to balance sending too many queries to the
+        // contacts API vs giving a feeling of responsiveness for the reaction
+        // of the autocomplete. This balance may be affected by the total number
+        // of contacts.
+        this.timedQueryId = setTimeout(this.runQuery, 350);
+      }
+    },
+
+
+    /**
+     * The owner of this autocomplete instance should set this value if they
+     * have extra spacing around the input that should be maintained for better
+     * visual display, and to avoid small scroll janks on initial positioning
+     * of the input element if it needs to be scrolled to the top of the view.
+     */
+    verticalSpace: 0,
+
+    /**
+     * This method can be overriden by the owner of the autocomplete to return
+     * existing entries for the given input node. The result should be an
+     * array of objects that have "name" and "address" properties.
+     * @param  {Element} inputNode The input element that is receiving the
+     * autocomplete.
+     * @return {Array}   Array of existing entries.
+     */
+    getExistingEntries: function(inputNode) {
+      return [];
+    },
+
+    /**
+     * Runs the query to find matches from the dataSource.
+     */
+    runQuery: function() {
+      this.timedQueryId = 0;
+
+      this.dataSource(this.query).then((result) => {
+        // Protect against old or out of date queries
+        if (result.query !== this.query) {
+          return;
+        }
+
+        var existingEntries = this.getExistingEntries(this.inputNode);
+
+        var contacts = result.contacts;
+
+        this.list.innerHTML = '';
+        this.topMatch = null;
+
+        var queryRegExp = new RegExp(this.query, 'i');
+
+        contacts.forEach((contact) => {
+          // Only care about contacts with an email address(es).
+          var emails = contact.email;
+          if (!emails || !emails.length) {
+            return;
+          }
+
+          // Contact name could be an array, just use the first one.
+          var name = contact.name;
+          if (Array.isArray(name)) {
+            name = name[0];
+          }
+
+          // Emails are an array, each one should get its own entry in the
+          // autocomplete.
+          emails.forEach((email) => {
+            var node = itemTemplate.cloneNode(true),
+                nameNode = node.querySelector('.cmp-autocomplete-name'),
+                emailNode = node.querySelector('.cmp-autocomplete-email'),
+                address = email.value;
+
+            // Do not add it if the email address is already an entry.
+            if (existingEntries.some(function(entry) {
+              return entry.address === address;
```

Comment:
I was too fancy here, extracted the some() call to a separate hasExisting var, then just that var in the if().